server/src/main/java/org/opensearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -134,6 +134,29 @@ public synchronized void markSeqNoAsPersisted(final long seqNo) {
         markSeqNo(seqNo, persistedCheckpoint, persistedSeqNo);
     }
 
+    /**
+     * Updates the processed checkpoint with the provided sequence number if segment replication is enabled
+     *
+     * @param seqNo the sequence number to mark as processed
+     */
+    public synchronized void segrepMarkSeqNoAsProcessed(final long seqNo) {
+        markSegrepSeqNo(seqNo, processedCheckpoint, persistedCheckpoint);
+    }
+
+    @SuppressForbidden(reason = "Object#notifyAll")
+    private void markSegrepSeqNo(final long seqNo, final AtomicLong processedCheckpoint, final AtomicLong persistedCheckpoint) {
+        assert Thread.holdsLock(this);
+        advanceMaxSeqNo(seqNo);
+        if ((seqNo > persistedCheckpoint.get()) || (seqNo <= processedCheckpoint.get())) {
+            return;
+        }
```

Comment:
A fair amount of this code is duplicated from the existing `markSeqNo` method. Since the first check is simply a bounds check, we can write a `shouldUpdateSeqNo` method that does this and returns a boolean to decide if further processing is necessary. This can be invoked by wrapper methods that perform the update by `compareAndSet` (as below) or via `updateCheckpoint` (as with the original method)